File: hadron-desktop/src-tauri/src/commands/jira.rs
//! JIRA integration commands — CRUD, search, linking, and deep analysis.

use super::common::DbState;
use crate::database::{Analysis, JiraLink};
use crate::jira_service;
use serde::Deserialize;
use std::sync::Arc;

// ============================================================================
// JIRA Core Commands (migrated from commands_legacy.rs)
// ============================================================================

/// Test JIRA connection
#[tauri::command]
pub async fn test_jira_connection(
    base_url: String,
    email: String,
    api_token: String,
) -> Result<jira_service::JiraTestResponse, String> {
    log::debug!("cmd: test_jira_connection");
    log::info!("Testing JIRA connection");
    jira_service::test_jira_connection(base_url, email, api_token).await
}

/// List JIRA projects for autocomplete
#[tauri::command]
pub async fn list_jira_projects(
    base_url: String,
    email: String,
    api_token: String,
) -> Result<Vec<jira_service::JiraProjectInfo>, String> {
    log::debug!("cmd: list_jira_projects");
    log::info!("Listing JIRA projects");
    jira_service::list_jira_projects(base_url, email, api_token).await
}

/// Create JIRA ticket from crash analysis
#[tauri::command]
pub async fn create_jira_ticket(
    base_url: String,
    email: String,
    api_token: String,
    project_key: String,
    issue_type: String,
    ticket: jira_service::JiraTicketRequest,
) -> Result<jira_service::JiraCreateResponse, String> {
    log::debug!("cmd: create_jira_ticket");
    log::info!("Creating JIRA ticket");
    jira_service::create_jira_ticket(base_url, email, api_token, project_key, issue_type, ticket)
        .await
}

/// Search JIRA issues using JQL
#[tauri::command]
pub async fn search_jira_issues(
    base_url: String,
    email: String,
    api_token: String,
    jql: String,
    max_results: i32,
    include_comments: bool,
) -> Result<jira_service::JiraSearchResponse, String> {
    log::debug!("cmd: search_jira_issues");
    log::info!("Searching JIRA issues with JQL");
    jira_service::search_jira_issues(base_url, email, api_token, jql, max_results, include_comments)
        .await
}

/// Post a comment to a JIRA issue
#[tauri::command]
pub async fn post_jira_comment(
    base_url: String,
    email: String,
    api_token: String,
    issue_key: String,
    comment_body: String,
) -> Result<(), String> {
    log::debug!("cmd: post_jira_comment");
    log::info!("Posting comment to JIRA issue {}", issue_key);
    jira_service::post_jira_comment(&base_url, &email, &api_token, &issue_key, &comment_body).await
}

/// Fetch the next page of JIRA issues using a cursor token
#[tauri::command]
pub async fn search_jira_issues_next_page(
    base_url: String,
    email: String,
    api_token: String,
    jql: String,
    max_results: i32,
    include_comments: bool,
    next_page_token: String,
) -> Result<jira_service::JiraSearchResponse, String> {
    log::debug!("cmd: search_jira_issues_next_page");
    jira_service::search_jira_issues_page_cursor(
        base_url,
        email,
        api_token,
        jql,
        Some(next_page_token),
        max_results,
        include_comments,
    )
    .await
}

/// Deep JIRA analysis — JIRA-specific prompt + structured JSON output stored in DB
#[tauri::command]
pub async fn analyze_jira_ticket_deep(
    request: crate::jira_deep_analysis::JiraDeepRequest,
    db: DbState<'_>,
) -> Result<serde_json::Value, String> {
    log::debug!("cmd: analyze_jira_ticket_deep key={}", request.jira_key);

    // Clone fields needed after request is moved into run_jira_deep_analysis
    let jira_key = request.jira_key.clone();
    let model = request.model.clone();
    let provider = request.provider.clone();

    let result = crate::jira_deep_analysis::run_jira_deep_analysis(request).await?;

    // Persist to DB using the existing analyses table
    let db = Arc::clone(&db);
    let result_for_db = result.clone();
    let jira_key_db = jira_key.clone();
    let model_db = model.clone();
    let provider_db = provider.clone();

    let analysis_id = tauri::async_runtime::spawn_blocking(move || {
        use crate::database::Analysis;

        // Remove any previous jira_deep analysis for this ticket to avoid duplicates
        db.delete_analyses_by_filename_and_type(&jira_key_db, "jira_deep")
            .map_err(|e| format!("Database error: {}", e))?;

        let full_data = serde_json::to_string(&result_for_db)
            .map_err(|e| format!("Serialization error: {}", e))?;

        let suggested_fixes_json = serde_json::to_string(
            &result_for_db
                .recommended_actions
                .iter()
                .map(|a| format!("[{}] {}", a.priority, a.action))
                .collect::<Vec<_>>(),
        )
        .unwrap_or_else(|_| "[]".to_string());

        let now = chrono::Utc::now().to_rfc3339();

        let analysis = Analysis {
            id: 0,
            filename: jira_key_db,
            file_size_kb: 0.0,
            error_type: result_for_db.technical.error_type.clone(),
            error_message: Some(result_for_db.plain_summary.clone()),
            severity: result_for_db.technical.severity_estimate.to_uppercase(),
            component: None,
            stack_trace: None,
            root_cause: result_for_db.technical.root_cause.clone(),
            suggested_fixes: suggested_fixes_json,
            confidence: Some(result_for_db.technical.confidence.clone()),
            analyzed_at: now,
            ai_model: model_db,
            ai_provider: Some(provider_db),
            tokens_used: 0,
            cost: 0.0,
            was_truncated: false,
            full_data: Some(full_data),
            is_favorite: false,
            last_viewed_at: None,
            view_count: 0,
            analysis_duration_ms: None,
            analysis_type: "jira_deep".to_string(),
        };

        db.insert_analysis(&analysis)
            .map_err(|e| format!("Database error: {}", e))
    })
    .await
    .map_err(|e| format!("Task error: {}", e))??;

    Ok(serde_json::json!({
        "id": analysis_id,
        "result": result,
    }))
}

// ============================================================================
// JIRA Ticket Linking Commands (migrated from commands_legacy.rs)
// ============================================================================

/// Link request from frontend
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LinkJiraTicketRequest {
    pub analysis_id: i64,
    pub jira_key: String,
    pub jira_url: Option<String>,
    pub jira_summary: Option<String>,
    pub jira_status: Option<String>,
    pub jira_priority: Option<String>,
    pub link_type: Option<String>,
    pub notes: Option<String>,
}

/// Link a JIRA ticket to an analysis
#[tauri::command]
pub async fn link_jira_to_analysis(
    request: LinkJiraTicketRequest,
    db: DbState<'_>,
) -> Result<JiraLink, String> {
    log::debug!("cmd: link_jira_to_analysis");
    log::info!(
        "Linking JIRA {} to analysis {}",
        request.jira_key,
        request.analysis_id
    );

    let db_clone = Arc::clone(&db);
    let analysis_id = request.analysis_id;

    tauri::async_runtime::spawn_blocking(move || {
        let link_type = request.link_type.as_deref().unwrap_or("related");

        db_clone
            .link_jira_ticket(
                analysis_id,
                &request.jira_key,
                request.jira_url.as_deref(),
                request.jira_summary.as_deref(),
                request.jira_status.as_deref(),
                request.jira_priority.as_deref(),
                link_type,
                request.notes.as_deref(),
            )
            .map_err(|e| format!("Failed to link JIRA ticket: {}", e))?;

        db_clone
            .get_jira_links_for_analysis(analysis_id)
            .map_err(|e| format!("Failed to get link: {}", e))?
            .into_iter()
            .find(|l| l.jira_key == request.jira_key)
            .ok_or_else(|| "Link not found after creation".to_string())
    })
    .await
    .map_err(|e| format!("Task error: {}", e))?
}

/// Unlink a JIRA ticket from an analysis
#[tauri::command]
pub async fn unlink_jira_from_analysis(
    analysis_id: i64,
    jira_key: String,
    db: DbState<'_>,
) -> Result<bool, String> {
    log::debug!("cmd: unlink_jira_from_analysis");
    log::info!("Unlinking JIRA {} from analysis {}", jira_key, analysis_id);

    let db_clone = Arc::clone(&db);

    tauri::async_runtime::spawn_blocking(move || db_clone.unlink_jira_ticket(analysis_id, &jira_key))
        .await
        .map_err(|e| format!("Task error: {}", e))?
        .map_err(|e| format!("Failed to unlink JIRA ticket: {}", e))
}

/// Get all JIRA links for an analysis
#[tauri::command]
pub async fn get_jira_links_for_analysis(
    analysis_id: i64,
    db: DbState<'_>,
) -> Result<Vec<JiraLink>, String> {
    log::debug!("cmd: get_jira_links_for_analysis");
    log::debug!("Getting JIRA links for analysis {}", analysis_id);

    let db_clone = Arc::clone(&db);

    tauri::async_runtime::spawn_blocking(move || db_clone.get_jira_links_for_analysis(analysis_id))
        .await
        .map_err(|e| format!("Task error: {}", e))?
        .map_err(|e| format!("Failed to get JIRA links: {}", e))
}

/// Get all analyses linked to a specific JIRA ticket
#[tauri::command]
pub async fn get_analyses_for_jira_ticket(
    jira_key: String,
    db: DbState<'_>,
) -> Result<Vec<(Analysis, JiraLink)>, String> {
    log::debug!("cmd: get_analyses_for_jira_ticket");
    log::debug!("Getting analyses linked to JIRA {}", jira_key);

    let db_clone = Arc::clone(&db);

    tauri::async_runtime::spawn_blocking(move || db_clone.get_analyses_for_jira_ticket(&jira_key))
        .await
        .map_err(|e| format!("Task error: {}", e))?
        .map_err(|e| format!("Failed to get analyses for JIRA ticket: {}", e))
}

/// Update JIRA ticket metadata in all links (e.g., after status change)
#[tauri::command]
pub async fn update_jira_link_metadata(
    jira_key: String,
    jira_summary: Option<String>,
    jira_status: Option<String>,
    jira_priority: Option<String>,
    db: DbState<'_>,
) -> Result<usize, String> {
    log::debug!("cmd: update_jira_link_metadata");
    log::info!("Updating JIRA {} metadata in links", jira_key);

    let db_clone = Arc::clone(&db);

    tauri::async_runtime::spawn_blocking(move || {
        db_clone.update_jira_link_metadata(
            &jira_key,
            jira_summary.as_deref(),
            jira_status.as_deref(),
            jira_priority.as_deref(),
        )
    })
    .await
    .map_err(|e| format!("Task error: {}", e))?
    .map_err(|e| format!("Failed to update JIRA link metadata: {}", e))
}

/// Count JIRA links for an analysis
#[tauri::command]
pub async fn count_jira_links_for_analysis(
    analysis_id: i64,
    db: DbState<'_>,
) -> Result<i64, String> {
    log::debug!("cmd: count_jira_links_for_analysis");
    let db_clone = Arc::clone(&db);

    tauri::async_runtime::spawn_blocking(move || db_clone.count_jira_links_for_analysis(analysis_id))
        .await
        .map_err(|e| format!("Task error: {}", e))?
        .map_err(|e| format!("Failed to count JIRA links: {}", e))
}

/// Get all JIRA links across all analyses (for sync service)
#[tauri::command]
pub async fn get_all_jira_links(db: DbState<'_>) -> Result<Vec<JiraLink>, String> {
    log::debug!("cmd: get_all_jira_links");
    log::debug!("Getting all JIRA links for sync");

    let db_clone = Arc::clone(&db);

    tauri::async_runtime::spawn_blocking(move || db_clone.get_all_jira_links())
        .await
        .map_err(|e| format!("Task error: {}", e))?
        .map_err(|e| format!("Failed to get all JIRA links: {}", e))
}

File: hadron-electron/electron/ipc/jira.ts
import { IpcMain } from 'electron'
import { getDb } from '../database'
import log from 'electron-log'
import { readJiraCreds, jiraFetch } from '../services/jira-client'

export function registerJiraHandlers(ipcMain: IpcMain): void {
  // ──────────────────────────────────────────────────────────────────────────
  // HTTP API handlers
  // ──────────────────────────────────────────────────────────────────────────

  ipcMain.handle('test_jira_connection', async () => {
    try {
      const { baseUrl, email, apiToken } = readJiraCreds()
      const data = await jiraFetch(baseUrl, email, apiToken, '/rest/api/3/myself') as { displayName?: string; emailAddress?: string }
      return {
        success: true,
        message: `Connected as ${data.displayName ?? data.emailAddress ?? 'unknown user'}`,
      }
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err)
      log.warn('test_jira_connection failed:', message)
      return { success: false, message }
    }
  })

  ipcMain.handle('list_jira_projects', async () => {
    try {
      const { baseUrl, email, apiToken } = readJiraCreds()
      const data = await jiraFetch(baseUrl, email, apiToken, '/rest/api/3/project') as Array<{ id: string; key: string; name: string }>
      return data.map(p => ({ id: p.id, key: p.key, name: p.name }))
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err)
      log.warn('list_jira_projects failed:', message)
      throw new Error(message)
    }
  })

  ipcMain.handle('create_jira_ticket', async (_e, args: {
    projectKey: string
    summary: string
    description?: string
    issueType?: string
    priority?: string
    labels?: string[]
    [key: string]: unknown
  }) => {
    try {
      const { baseUrl, email, apiToken } = readJiraCreds()
      const body = {
        fields: {
          project: { key: args.projectKey },
          summary: args.summary,
          description: args.description
            ? {
                type: 'doc',
                version: 1,
                content: [{ type: 'paragraph', content: [{ type: 'text', text: args.description }] }],
              }
            : undefined,
          issuetype: { name: args.issueType ?? 'Bug' },
          ...(args.priority ? { priority: { name: args.priority } } : {}),
          ...(args.labels ? { labels: args.labels } : {}),
        },
      }
      const data = await jiraFetch(baseUrl, email, apiToken, '/rest/api/3/issue', {
        method: 'POST',
        body: JSON.stringify(body),
      }) as { key: string; id: string }
      return {
        key: data.key,
        id: data.id,
        url: `${baseUrl.replace(/\/$/, '')}/browse/${data.key}`,
      }
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err)
      log.warn('create_jira_ticket failed:', message)
      throw new Error(message)
    }
  })

  ipcMain.handle('search_jira_issues', async (_e, args: {
    jql: string
    maxResults?: number
    fields?: string[]
  }) => {
    try {
      const { baseUrl, email, apiToken } = readJiraCreds()
      const params = new URLSearchParams({
        jql: args.jql,
        maxResults: String(Math.min(args.maxResults ?? 50, 100)),
      })
      if (args.fields && args.fields.length > 0) {
        params.set('fields', args.fields.join(','))
      }
      return jiraFetch(baseUrl, email, apiToken, `/rest/api/3/issue/search?${params.toString()}`)
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err)
      log.warn('search_jira_issues failed:', message)
      throw new Error(message)
    }
  })

  ipcMain.handle('search_jira_issues_next_page', async (_e, args: {
    jql: string
    startAt: number
    maxResults?: number
    fields?: string[]
  }) => {
    try {
      const { baseUrl, email, apiToken } = readJiraCreds()
      const params = new URLSearchParams({
        jql: args.jql,
        startAt: String(args.startAt),
        maxResults: String(Math.min(args.maxResults ?? 50, 100)),
      })
      if (args.fields && args.fields.length > 0) {
        params.set('fields', args.fields.join(','))
      }
      return jiraFetch(baseUrl, email, apiToken, `/rest/api/3/issue/search?${params.toString()}`)
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err)
      log.warn('search_jira_issues_next_page failed:', message)
      throw new Error(message)
    }
  })

  ipcMain.handle('post_jira_comment', async (_e, args: {
    issueKey: string
    body: string
  }) => {
    try {
      const { baseUrl, email, apiToken } = readJiraCreds()
      const commentBody = {
        body: {
          type: 'doc',
          version: 1,
          content: [{ type: 'paragraph', content: [{ type: 'text', text: args.body }] }],
        },
      }
      return jiraFetch(baseUrl, email, apiToken, `/rest/api/3/issue/${encodeURIComponent(args.issueKey)}/comment`, {
        method: 'POST',
        body: JSON.stringify(commentBody),
      })
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err)
      log.warn('post_jira_comment failed:', message)
      throw new Error(message)
    }
  })

  // ──────────────────────────────────────────────────────────────────────────
  // DB linking handlers  (table: analysis_jira_links)
  // ──────────────────────────────────────────────────────────────────────────

  ipcMain.handle('link_jira_to_analysis', (_e, args: {
    analysisId: number
    jiraKey: string
    jiraUrl?: string
    jiraSummary?: string
    jiraStatus?: string
    jiraPriority?: string
    linkType?: string
    linkedBy?: string
    notes?: string
  }) => {
    const db = getDb()
    const now = new Date().toISOString()
    db.prepare(`
      INSERT OR REPLACE INTO analysis_jira_links
        (analysis_id, jira_key, jira_url, jira_summary, jira_status, jira_priority,
         link_type, linked_at, linked_by, notes)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(
      args.analysisId,
      args.jiraKey,
      args.jiraUrl ?? null,
      args.jiraSummary ?? null,
      args.jiraStatus ?? null,
      args.jiraPriority ?? null,
      args.linkType ?? 'related',
      now,
      args.linkedBy ?? null,
      args.notes ?? null,
    )
    return db.prepare(
      'SELECT * FROM analysis_jira_links WHERE analysis_id = ? AND jira_key = ?'
    ).get(args.analysisId, args.jiraKey)
  })

  ipcMain.handle('unlink_jira_from_analysis', (_e, args: {
    analysisId: number
    jiraKey: string
  }) => {
    const db = getDb()
    const result = db.prepare(
      'DELETE FROM analysis_jira_links WHERE analysis_id = ? AND jira_key = ?'
    ).run(args.analysisId, args.jiraKey)
    return result.changes > 0
  })

  ipcMain.handle('get_jira_links_for_analysis', (_e, args: { analysisId: number }) => {
    const db = getDb()
    return db.prepare(
      'SELECT * FROM analysis_jira_links WHERE analysis_id = ? ORDER BY linked_at DESC'
    ).all(args.analysisId)
  })

  ipcMain.handle('get_analyses_for_jira_ticket', (_e, args: { jiraKey: string }) => {
    const db = getDb()
    const rows = db.prepare(`
      SELECT a.*, l.id AS link_id, l.link_type, l.linked_at, l.linked_by, l.notes,
             l.jira_url, l.jira_summary, l.jira_status, l.jira_priority
      FROM analysis_jira_links l
      INNER JOIN analyses a ON a.id = l.analysis_id
      WHERE l.jira_key = ?
      ORDER BY l.linked_at DESC
    `).all(args.jiraKey) as Record<string, unknown>[]
    return rows.map(row => {
      const link = {
        id: row.link_id,
        analysis_id: row.id,
        jira_key: args.jiraKey,
        jira_url: row.jira_url,
        jira_summary: row.jira_summary,
        jira_status: row.jira_status,
        jira_priority: row.jira_priority,
        link_type: row.link_type,
        linked_at: row.linked_at,
        linked_by: row.linked_by,
        notes: row.notes,
      }
      const analysis = { ...row }
      delete analysis.link_id
      delete analysis.link_type
      delete analysis.linked_at
      delete analysis.linked_by
      delete analysis.notes
      delete analysis.jira_url
      delete analysis.jira_summary
      delete analysis.jira_status
      delete analysis.jira_priority
      return [analysis, link]
    })
  })

  ipcMain.handle('update_jira_link_metadata', (_e, args: {
    jiraKey: string
    jiraSummary?: string
    jiraStatus?: string
    jiraPriority?: string
  }) => {
    const db = getDb()
    const updates: string[] = []
    const params: unknown[] = []
    if (args.jiraSummary !== undefined) { updates.push('jira_summary = ?'); params.push(args.jiraSummary) }
    if (args.jiraStatus !== undefined) { updates.push('jira_status = ?'); params.push(args.jiraStatus) }
    if (args.jiraPriority !== undefined) { updates.push('jira_priority = ?'); params.push(args.jiraPriority) }
    if (updates.length === 0) return 0
    params.push(args.jiraKey)
    const result = db.prepare(
      `UPDATE analysis_jira_links SET ${updates.join(', ')} WHERE jira_key = ?`
    ).run(...params)
    return result.changes
  })

  ipcMain.handle('count_jira_links_for_analysis', (_e, args: { analysisId: number }) => {
    const db = getDb()
    const row = db.prepare(
      'SELECT COUNT(*) AS count FROM analysis_jira_links WHERE analysis_id = ?'
    ).get(args.analysisId) as { count: number }
    return row.count
  })

  ipcMain.handle('get_all_jira_links', () => {
    const db = getDb()
    return db.prepare(
      'SELECT * FROM analysis_jira_links ORDER BY linked_at DESC'
    ).all()
  })
}
